downloader.go

```diff
@@ -0,0 +1,51 @@
+//go:build !NODOWNLOAD
+
+package hugot
+
+import (
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	hfd "github.com/bodaay/HuggingFaceModelDownloader/hfdownloader"
+)
+
+type DownloadOptions struct {
+	AuthToken             string
+	SkipSha               bool
+	Branch                string
+	MaxRetries            int
+	RetryInterval         int
+	ConcurrentConnections int
+	Verbose               bool
+}
+
+func NewDownloadOptions() DownloadOptions {
+	d := DownloadOptions{}
+	d.Branch = "main"
+	d.MaxRetries = 5
+	d.RetryInterval = 5
+	d.ConcurrentConnections = 5
+	return d
+}
+
+func (s *Session) DownloadModel(modelName string, destination string, options DownloadOptions) (string, error) {
+	// todo: replicates code in hf downloader, function can be modified there
+	modelP := modelName
+	if strings.Contains(modelP, ":") {
+		modelP = strings.Split(modelName, ":")[0]
+	}
+	modelPath := path.Join(destination, strings.Replace(modelP, "/", "_", -1))
+
+	for i := 0; i < options.MaxRetries; i++ {
+		if err := hfd.DownloadModel(modelName, false, options.SkipSha, false, destination, options.Branch, options.ConcurrentConnections, options.AuthToken, !options.Verbose); err != nil {
+			fmt.Printf("Warning: attempt %d / %d failed, error: %s\n", i+1, options.MaxRetries, err)
+			time.Sleep(time.Duration(options.RetryInterval) * time.Second)
+			continue
+		}
+		fmt.Printf("\nDownload of %s completed successfully\n", modelName)
+		return modelPath, nil
+	}
+	return "", fmt.Errorf("failed to download %s after %d attempts", modelName, options.MaxRetries)
+}
```

Comment:
Might be good to check downloaded model actualy has the required files, e.g. tokenizer and onnx?

Reply:
done. It's a bit of a question whether we want to be strict, i.e. there must be a "model.onnx" file, or whether we try to load the first ".onnx" file we find

Reply:
Yeh difficult one, there is no real standard here, sometimes you get model and model quantized. Probs load if exactly one exists, else they must provide the file name as a param?